pkg/block/fetcher.go

```diff
@@ -410,3 +411,52 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
 		delete(metas, id)
 	}
 }
+
+// CompactedBlockFilter is a MetaFetcher filter that filters out blocks that have newer block
+// from two or more overlapping blocks that fully submatches the source blocks of the older blocks.
+type CompactedBlockFilter struct{}
+
+// NewCompactedBlockFilter creates CompactedBlockFilter.
+func NewCompactedBlockFilter() *CompactedBlockFilter {
+	return &CompactedBlockFilter{}
+}
+
+// Filter filters out blocks that filters out blocks that have newer block
+// from two or more overlapping blocks that fully submatches the source blocks of the older blocks.
+func (f *CompactedBlockFilter) Filter(metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) {
+	for id, meta := range metas {
+		sources := make([]ulid.ULID, len(metas))
+		for _, sourceID := range meta.Compaction.Sources {
+			if sourceID == id {
+				continue
+			}
+			sourceMeta := metas[sourceID]
+			if sourceMeta == nil {
+				break
+			}
+			sources = append(sources, sourceMeta.Compaction.Sources...)
+		}
+		sort.Slice(sources, func(i, j int) bool {
+			return sources[i].Compare(sources[j]) < 0
+		})
+		if areSlicesEqual(sources, meta.Compaction.Sources) {
+			for _, sourceID := range meta.Compaction.Sources {
+				if sourceID != id {
+					delete(metas, sourceID)
```

Comment:
Next to deleting, we probably want store the deleted metas and check if we are ready to delete those blocks from bucket (: